### PR TITLE
Migrate `font` custom properties from v11 to v12

### LINKED ISSUE #10532: Run `font` migration on next branch

### PR DESCRIPTION
**Pull request by laurkim**
### WHY are these changes introduced?

Resolves #10532.
Migrates `font` custom properties.
Updates body and heading variants in `Text` components to use text composite tokens where applicable.

> [!Important]
> Font migration needed to be run in 4 sequential steps due to overlapping `font-size` token names. 

## Step 1

```script
npx @shopify/polaris-migrator@0.0.0-snapshot-release-20230928163203 \
v12-styles-replace-custom-property-font \
'./**/*.scss' \
--step=1
```

```sh
# Stage all migrated files
git add .
# Format staged files only
git diff --staged --name-only | xargs npx prettier --write
# Commit automated migrations
git commit -m "Migrate `font` custom properties from Polaris v11 to v12 — step [1/2/3/4]“
```

### Post-migration validation
<details>
  <summary>[v12] Check RegExp for hardcoded <code>font</code> custom properties in step 1 across all file types
  </summary>
 
```
(?:--p-font-size-70-experimental|--p-font-size-80-experimental|--p-font-size-100|--p-font-size-700|--p-font-line-height-075-experimental|--p-font-line-height-1|--p-font-line-height-2|--p-font-line-height-3|--p-font-line-height-4|--p-font-line-height-5|--p-font-line-height-6|--p-font-line-height-7)(?![\w-])
```
</details>

## Step 2

```script
# Run migration
npx @shopify/polaris-migrator@0.0.0-snapshot-release-20230928163203 \
v12-styles-replace-custom-property-font \
'./**/*.scss' \
--step=2

# Repeat process in step 1 to commit automated migrations
```

### Post-migration validation
<details>
  <summary>[v12] Check RegExp for hardcoded <code>font</code> custom properties in step 2 across all file types
  </summary>
 
```
(?:--p-font-size-500|--p-font-size-600)(?![\w-])
```
</details>

## Step 3

```script
# Run migration
npx @shopify/polaris-migrator@0.0.0-snapshot-release-20230928163203 \
v12-styles-replace-custom-property-font \
'./**/*.scss' \
--step=3

#  Repeat process in step 1 to commit automated migrations
```

### Post-migration validation
<details>
  <summary>[v12] Check RegExp for hardcoded <code>font</code> custom properties in step 3 across all file types
  </summary>
</details>

```
(?:--p-font-size-300|--p-font-size-400)(?![\w-])
```
</details>

## Step 4

```script
# Run migration
npx @shopify/polaris-migrator@0.0.0-snapshot-release-20230928163203 \
v12-styles-replace-custom-property-font \
'./**/*.scss' \
--step=4

# Repeat process in step 1 to commit automated migrations
```

### Post-migration validation
<details>
  <summary>[v12] Check RegExp for hardcoded <code>font</code> custom properties in step 4 across all file types
  </summary>
 
```
(?:--p-font-size-75|--p-font-size-200)(?![\w-])
```
</details>

#### Replacement maps

| Deprecated CSS Custom Property   | Replacement Value |
| ---------------- | ----------------- |
| `--p-font-size-70-experimental` | `--p-font-size-275` |
| `--p-font-size-75` | `--p-font-size-300` |
| `--p-font-size-80-experimental` | `--p-font-size-325` |
| `--p-font-size-100` | `--p-font-size-350` |
| `--p-font-size-200` | `--p-font-size-400` |
| `--p-font-size-300` | `--p-font-size-500` |
| `--p-font-size-400` | `--p-font-size-600` |
| `--p-font-size-500` | `--p-font-size-750` |
| `--p-font-size-600` | `--p-font-size-900` |
| `--p-font-size-700` | `--p-font-size-1000` |
| `--p-font-line-height-075-experimental` | `--p-font-line-height-300` | 
| `--p-font-line-height-1` | `--p-font-line-height-400` | 
| `--p-font-line-height-2` | `--p-font-line-height-500` | 
| `--p-font-line-height-3` | `--p-font-line-height-600` | 
| `--p-font-line-height-4` | `--p-font-line-height-700` | 
| `--p-font-line-height-5` | `--p-font-line-height-800` | 
| `--p-font-line-height-6` | `--p-font-line-height-1000` | 
| `--p-font-line-height-7` | `--p-font-line-height-1200` | 

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

File: documentation/guides/legacy-polaris-v8-public-api.scss
Comment (laurkim) [inline documentation/guides/legacy-polaris-v8-public-api.scss:1050]: These are changes from the previous border migration.
I inadvertently updated this file in the border migration PR so I just reverted the changes in this PR.

**Review by lgriffee — APPROVED** [review]
💯🚢✨🎉

**Review by lgriffee — APPROVED** [review]
❤️

**Review by bernardojoaogarcia — COMMENTED** [review]
Could we verify the font size and line height for heading-xl, 2xl & 3xl. I'm not sure they're correct in this PR

**Comment by bernardojoaogarcia** [comment]
**heading-3xl**
```
font-size-900 (36px)
font-line-height-1200 (48px)
font-weight-bold
font-letter-spacing-densest
font-family-sans
```

**heading-2xl**
```
font-size-750 (30px)
font-line-height-1000 (40px)
font-weight-bold
font-letter-spacing-denser
font-family-sans

```
**heading-xl**
```
font-size-600 (24px)
font-line-height-800 (32px)
font-weight-bold
font-letter-spacing-dense
font-family-sans
```

**heading-lg**
```
font-size-500 (20px)
font-line-height-600 (24px)
font-weight-semibold
font-letter-spacing-dense
font-family-sans
```

File: polaris-react/src/components/Text/Text.scss
Comment (laurkim) [inline polaris-react/src/components/Text/Text.scss:87]: I only added the headingLg composite tokens for `breakpoints-md-up` since responsiveness was out of scope when these tokens were created.

File: polaris-react/src/components/Text/Text.scss
Comment (laurkim) [inline polaris-react/src/components/Text/Text.scss:98]: I only added the headingXl composite tokens for `breakpoints-md-up` since responsiveness was out of scope when these tokens were created.

File: polaris-react/src/components/Text/Text.scss
Comment (laurkim) [inline polaris-react/src/components/Text/Text.scss:109]: I only added the heading2xl composite tokens for `breakpoints-md-up` since responsiveness was out of scope when these tokens were created.

File: polaris-react/src/components/Text/Text.scss
Comment (laurkim) [inline polaris-react/src/components/Text/Text.scss:120]: I only added the heading3xl composite tokens for `breakpoints-md-up` since responsiveness was out of scope when these tokens were created.

**Review by lgriffee — APPROVED** [review]
✨🙌🏻

**Review by sarahill — APPROVED** [review]

**Comment by bernardojoaogarcia** [comment]
@sarahill aren't we supposed to make an exception for the primary buttons `font-weight`? Instead of `font-weight-medium` isn't it supposed to be `font-weight-semibold`? Every other button is medium.

**Comment by laurkim** [comment]
> @sarahill aren't we supposed to make an exception for the primary buttons `font-weight`? Instead of `font-weight-medium` isn't it supposed to be `font-weight-semibold`? Every other button is medium.

@bernardojoaogarcia @sarahill I updated Button primary variant to use `font-weight-semibold`. I'm going to merge this since it has approvals to unblock other migrations on the `next` branch but if we decide to change primary Button font weight from the changes on this branch, let me know and I can fix forward 👍 

28-58-12av1-vx71n

**Comment by bernardojoaogarcia** [comment]
Thanks! @laurkim

**Pull request merged**
